asset/boltdb_manager.go

```diff
@@ -0,0 +1,135 @@
+// provides the serialization guarantee that the asset contract specifies.
+// We rely on long-lived BoltDB transactions during Get to provide this
+// mechanism for blocking.
+type BoltDBAssetManager struct {
+	// CacheDir specifies the location of local asset storage.
+	LocalStorage string
+
+	// DB is the BoltDB
+	DB *bolt.DB
+
+	Fetcher
+	Expander
+	Verifier
+}
+
+// Get opens a read-write transaction to BoltDB, causing subsequent calls to
+// Get to block. During this transaction, we attempt to determine if the asset
+// is installed by querying BoltDB for the asset's SHA (which we use as an ID).
+//
+// If a value is returned, we return the deserialized asset stored in BoltDB.
+// If deserialization fails, we assume there is some level of corruption and
+// attempt to re-install the asset.
+//
+// If a value is not returned, the asset is not installed or not installed
+// correctly. We then proceed to attempt asset installation.
+func (b *BoltDBAssetManager) Get(asset *types.Asset) (*RuntimeAsset, error) {
+	var localAsset *RuntimeAsset
+	key := []byte(asset.Sha512)
+
+	// This is racey, but the udpate transaction that comes over this view
+	// will cause all other update transactions to block. We always want
+	// to allow this Get() to return if the asset is already installed.
+	if err := b.DB.View(func(tx *bolt.Tx) error {
+		// If the key exists, the bucket should already exist.
+		bucket := tx.Bucket(assetBucketName)
+		if bucket == nil {
+			return nil
+		}
+
+		value := bucket.Get(key)
+		if value != nil {
+			// deserialize asset
+			if err := json.Unmarshal(value, &localAsset); err == nil {
+				return nil
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	// Check to see if the view was successful.
+	if localAsset != nil {
+		return localAsset, nil
+	}
+
+	if err := b.DB.Update(func(tx *bolt.Tx) error {
+		bucket, err := tx.CreateBucketIfNotExists(assetBucketName)
+		if err != nil {
+			return err
+		}
+
+		// Though we've already attempted to do this, it's possible that a previous
+		// call completed installation of the asset while this transaction
+		// was blocked on serialization. Re-attempt to get the key in case that is
+		// what happened.
+		value := bucket.Get(key)
+		if value != nil {
+			// deserialize asset
+			if err := json.Unmarshal(value, &localAsset); err == nil {
+				return nil
+			}
+		}
+
+		// install the asset
+		tmpFile, err := b.Fetch(asset.URL)
+		if err != nil {
+			return err
+		}
+		defer tmpFile.Close()
+		defer os.Remove(tmpFile.Name())
+
+		// verify
+		if err := b.Verify(tmpFile, asset.Sha512); err != nil {
+			return err
+		}
+
+		// expand
+		assetPath := filepath.Join(b.LocalStorage, asset.Sha512)
+		if err := b.Expand(tmpFile, assetPath); err != nil {
+			return err
+		}
+
+		localAsset = &RuntimeAsset{
+			Path: assetPath,
+		}
+
+		assetJSON, err := json.Marshal(localAsset)
+		if err != nil {
+			return err
```

Comment:
If marshaling assets results in an error, then this will cause the asset downloader to infinitely loop, downloading assets over and over, erroring on marshal.

Marshaling an asset probably won't result in an error - the RuntimeAsset would need to contain an incompatible type for that to happen.

We should panic if the error is non-nil, and add a test that ensures that RuntimeAssets can be marshaled without error.

Reply:
Oh yeah, that's a good point. Panicking instead.